### PR TITLE
feat: enhance autoscaling peeking logic

### PR DESCRIPTION
**Pull request by akashjkhamkar**
### Problem statement -
We periodically wake up pods (scale them back to 1 from 0) to check if any messages are pending the buffer or not.
This causes avalanche of k8s events in the cluster if the overall numaflow pod count is too high in the cluster.

### Solution -
We can simply query the deamon to check if any messages are pending or not, then we can decide whether to scale back the pod or not.

### Changes made -
vertex controller
(scaling.go)
If the replicas == 0, and vertex is not a source vertex; then check if that vertex has any pending mesages or not.
If yes, then scale it back , else do nothing.

SS
Screenshot 2023-12-30 at 2 05 15 AM

### CODE REVIEW

File: pkg/daemon/server/service/pipeline_metrics_query.go [hunk 1]
```diff
@@ -144,17 +144,30 @@ func (ps *PipelineMetadataQuery) GetVertexMetrics(ctx context.Context, req *daem
 	if abstractVertex.IsASource() {
 		bufferList = append(bufferList, req.GetVertex())
 	}
-	partitionPendingInfo := ps.getPending(ctx, req)
```
Comment (whynowy) [inline pkg/daemon/server/service/pipeline_metrics_query.go:147]: The implementation of this function should not be changed. The pending information returned in this implementation is a set of average values in the past period of times, and we use one of the average values to compute desired replicas. 

To find the instant buffer pending number, we can use GetPipelineBuffer().

File: Makefile [hunk 2]
```diff
@@ -129,6 +129,13 @@ test-%:
 	go test -v -timeout 15m -count 1 --tags test -p 1 ./test/$*
 	$(MAKE) cleanup-e2e
 	
+delete-and-install:
```
[the remark this answers is not in the document]
Comment (akashjkhamkar) [reply Makefile:132]: I couldn't find a make command which will delete and reinstall all control plane components, this was helpful in developing and testing controllers quickly.

Let me know if a similar command exists, i'll remove this 👍

File: pkg/daemon/server/service/pipeline_metrics_query.go (hunk 1 above)
Comment (akashjkhamkar) [reply pkg/daemon/server/service/pipeline_metrics_query.go]: Reverted deamon code 👍 

Using GetPipelineBuffer() to fetch relevant buffers of the vertex now

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 3]
```diff
@@ -223,6 +215,30 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
 		s.daemonClientsCache.Add(pl.GetDaemonServiceURL(), daemonClient)
 	}
 
+	if vertex.Status.Replicas == 0 { // Was scaled to 0
+		// Non source udfs don't need to scale up and peek to check the pending messages
+		// We will only scale them up if they have any pending messages
+		if !vertex.IsASource() {
+			totalPending, err := getPendingUpstreamMessageCount(ctx, daemonClient, pl, vertex)
+			if err != nil {
+				return fmt.Errorf("Failed to get the pending upstream messages: %s", err.Error())
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:224]:
```suggestion
				return fmt.Errorf("failed to get the ISB pending messages: %w", err)
```

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 4]
```diff
@@ -498,3 +514,24 @@ func (s *Scaler) patchVertexReplicas(ctx context.Context, vertex *dfv1.Vertex, d
 func KeyOfVertex(vertex dfv1.Vertex) string {
 	return fmt.Sprintf("%s/%s", vertex.Namespace, vertex.Name)
 }
+
+func getPendingUpstreamMessageCount(
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:518]: `getISBPendingMessageCount`

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 5]
```diff
@@ -223,6 +215,30 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
 		s.daemonClientsCache.Add(pl.GetDaemonServiceURL(), daemonClient)
 	}
 
+	if vertex.Status.Replicas == 0 { // Was scaled to 0
+		// Non source udfs don't need to scale up and peek to check the pending messages
+		// We will only scale them up if they have any pending messages
+		if !vertex.IsASource() {
+			totalPending, err := getPendingUpstreamMessageCount(ctx, daemonClient, pl, vertex)
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:222]: There's already some code to query the buffer information (to get `partitionBufferLengths`, etc), could you combine them? e.g.

```golang
func getBufferInfos() (xxxx, ...) {
  // return totalPending, partitionBufferLengths, partitionAvailableBufferLengths, error
}
```

`targetAvailableBufferLength` is not in-use, can be removed.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (whynowy) [reply Makefile:132]: Maybe move them into `make image`, and `make e2eapi-image`, under

```
ifdef IMAGE_IMPORT_CMD
        $(IMAGE_IMPORT_CMD) $(IMAGE_NAMESPACE)/$(BINARY_NAME):$(VERSION)
        # here.
endif
```

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 6]
```diff
@@ -223,6 +215,30 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
...
+				return fmt.Errorf("Failed to get the pending upstream messages: %s", err.Error())
+			}
+
+			if totalPending <= 0 {
+				log.Debugf("Vertex %s doesn't have any pending messages, skipping scaling back to 1", vertex.Name)
+				return nil
+			}
+		}
+
+		if secondsSinceLastScale >= float64(vertex.Spec.Scale.GetZeroReplicaSleepSeconds()) {
```
[the remark this answers is not in the document]
Comment (akashjkhamkar) [reply pkg/reconciler/vertex/scaling/scaling.go:238]: updated the doc for `zeroReplicaSleepSeconds`

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (akashjkhamkar) [reply Makefile:132]: I think `image` and `e2eapi-image` should only be dedicated for creating new image.

For deleting, we should have a separate command like  `delete-numaflow-controller-ux-webhook`

```
delete-numaflow-controller-ux-webhook:
	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=controller-manager,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-ux,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-webhook,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
	kubectl -n numaflow-system delete po e2e-api-pod --ignore-not-found=true
```

and for creating new image + deleting old image pods
```
reinstall-numaflow-controller-ux-webhook:
	$(MAKE) cleanup-e2e
	$(MAKE) delete-numaflow-controller-ux-webhook
```

[the remark this answers is not in the document]
Comment (akashjkhamkar) [reply Makefile:132]: Let me know how it sounds

File: pkg/reconciler/vertex/scaling/scaling.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (akashjkhamkar) [reply pkg/reconciler/vertex/scaling/scaling.go:238]: got it, will bypass `if secondsSinceLastScale >= float64(vertex.Spec.Scale.GetZeroReplicaSleepSeconds()) {
akashjkhamkar marked this conversation as resolved.
` if the vertex is non source and there are pending messages

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (whynowy) [reply Makefile]: `image` and `e2eapi-image` are also import images if a local cluster is used. If it's local cluster, there's no reason that you need to restart the pods.

If you look at your changes, all the pod deletion behaviors follow an `image` operation.

File: docs/user-guide/reference/autoscaling.md [hunk 7]
```diff
@@ -53,8 +53,8 @@ spec:
   start processing, because the autoscaling algorithm will divide the TPS by the number of pods even if the pod is not `Running`.
 - `scaleDownCooldownSeconds` - After a scaling operation, how many seconds to wait for the same vertex, if the follow-up
   operation is a scaling down, defaults to `90`.
-- `zeroReplicaSleepSeconds` - How many seconds it will wait after scaling down to `0`, defaults to `120`.
-  Numaflow autoscaler periodically scales up a vertex pod to "peek" the incoming data, this is the period of time to wait before peeking.
+- `zeroReplicaSleepSeconds` - How many seconds it will wait after scaling a source pod down to `0`, defaults to `120`.
```
Comment (whynowy) [inline docs/user-guide/reference/autoscaling.md:56]:
```suggestion
- `zeroReplicaSleepSeconds` - How many seconds it will wait after scaling a source vertex replicas down to `0`, defaults to `120`.
```

File: docs/user-guide/reference/autoscaling.md [hunk 8]
```diff
@@ -53,8 +53,8 @@ spec:
...
+  Numaflow autoscaler periodically scales up a source pod to "peek" the incoming data, this is the period of time to wait before peeking.
```
Comment (whynowy) [inline docs/user-guide/reference/autoscaling.md:57]:
```suggestion
  Numaflow autoscaler periodically scales up a source vertex pod to "peek" the incoming data, this is the period of time to wait before peeking.
```

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 9]
```diff
@@ -223,6 +215,37 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
 		s.daemonClientsCache.Add(pl.GetDaemonServiceURL(), daemonClient)
 	}
 
+	partitionBufferLengths, partitionAvailableBufferLengths, _, totalCurrentPending, err := s.getBufferInfos(ctx, key, daemonClient, pl, vertex)
+	if err != nil {
+		log.Debugf("error while fetching buffer info, %w", err)
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:220]: No need to add `Debugf`, `return fmt.Errorf("xxxxxx, %w", err)`

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 10]
```diff
@@ -498,3 +501,44 @@ func (s *Scaler) patchVertexReplicas(ctx context.Context, vertex *dfv1.Vertex, d
...
+				}
+
+				partitionBufferLengths = append(partitionBufferLengths, int64(float64(bInfo.GetBufferLength())*bInfo.GetBufferUsageLimit()))
+				partitionAvailableBufferLengths = append(partitionAvailableBufferLengths, int64(float64(bInfo.GetBufferLength())*float64(vertex.Spec.Scale.GetTargetBufferAvailability())/100))
+				totalBufferLength += int64(float64(*bInfo.BufferLength) * *bInfo.BufferUsageLimit)
+				totalCurrentPending += *bInfo.PendingCount
+			}
+		}
+
+		_ = s.vertexMetricsCache.Add(key+"/length", totalBufferLength)
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:540]: Return the `totalBufferLength`, and then add it to the cache in `scaleOneVertex()`, make it closer to adding `key+"/pending"` to the cache, so that it would be easy to read the code. In that case, change this to a standalone function.

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 11]
```diff
@@ -498,3 +501,44 @@ func (s *Scaler) patchVertexReplicas(ctx context.Context, vertex *dfv1.Vertex, d
...
+	partitionAvailableBufferLengths[]int64,
+	totalBufferLength int64,
+	totalCurrentPending int64,
+	err error,
+) {
+	partitionBufferLengths = make([]int64, 0)
+	partitionAvailableBufferLengths = make([]int64, 0)
+	totalBufferLength = int64(0)
+	totalCurrentPending = int64(0)
+	if !vertex.IsASource() { // Only non-source vertex has buffer to read
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:522]: No need to add this check, since `OwnedBuffers()` of source vertex has 0 item.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (akashjkhamkar) [reply Makefile]: agreed that `make image` will import the image in the cluster, but we still require pods to be killed, so that we can run the new image.

My goal was to: 
- create a new image
- import into the cluster
- create pods using this new image

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 12]
```diff
@@ -261,27 +284,8 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
...
-				partitionBufferLengths = append(partitionBufferLengths, int64(float64(bInfo.GetBufferLength())*bInfo.GetBufferUsageLimit()))
-				partitionAvailableBufferLengths = append(partitionAvailableBufferLengths, int64(float64(bInfo.GetBufferLength())*float64(vertex.Spec.Scale.GetTargetBufferAvailability())/100))
-				totalBufferLength += int64(float64(*bInfo.BufferLength) * *bInfo.BufferUsageLimit)
-				targetAvailableBufferLength += int64(float64(*bInfo.BufferLength) * float64(vertex.Spec.Scale.GetTargetBufferAvailability()) / 100)
-			}
-		}
-		// Add processing rate information to cache for back pressure calculation
-		_ = s.vertexMetricsCache.Add(key+"/length", totalBufferLength)
-	}
+	_ = s.vertexMetricsCache.Add(key+"/length", totalBufferLength)
```
Comment (whynowy) [inline pkg/reconciler/vertex/scaling/scaling.go:287]: Non-source vertex check.

File: Makefile [hunk 13]
```diff
@@ -120,15 +120,21 @@ test-sideinputs-e2e:
...
 	go generate $(shell find ./test/$* -name '*.go')
 	go test -v -timeout 15m -count 1 --tags test -p 1 ./test/$*
 	$(MAKE) cleanup-e2e
-	
+
+reinstall-numaflow-controller-ux-webhook:
+	$(MAKE) image
+	$(MAKE) delete-numaflow-controller-ux-webhook
+
+delete-numaflow-controller-ux-webhook:
```
Comment (whynowy) [inline Makefile:133]: `restart-control-plane-components`

File: Makefile [hunk 14]
```diff
@@ -120,15 +120,21 @@ test-sideinputs-e2e:
...
+	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=controller-manager,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
+	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-ux,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
+	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-webhook,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
+	kubectl -n numaflow-system delete po e2e-api-pod --ignore-not-found=true
```
Comment (whynowy) [inline Makefile:137]: This is only for e2e testing purpose, should not be part of this. Can make it remain in the `test-%`.

File: Makefile [hunk 15]
```diff
@@ -120,15 +120,21 @@ test-sideinputs-e2e:
...
-	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-webhook,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
-	kubectl -n numaflow-system delete po e2e-api-pod --ignore-not-found=true
+	$(MAKE) delete-numaflow-controller-ux-webhook
 	cat test/manifests/e2e-api-pod.yaml |  sed 's@quay.io/numaproj/@$(IMAGE_NAMESPACE)/@' | sed 's/:latest/:$(VERSION)/' | kubectl -n numaflow-system apply -f -
 	go generate $(shell find ./test/$* -name '*.go')
 	go test -v -timeout 15m -count 1 --tags test -p 1 ./test/$*
 	$(MAKE) cleanup-e2e
-	
+
+reinstall-numaflow-controller-ux-webhook:
```
Comment (whynowy) [inline Makefile:129]: Nobody likes to use such a long make command. Use `image-restart`?

File: Makefile [hunk 16]
```diff
@@ -120,15 +120,21 @@ test-sideinputs-e2e:
...
 	kubectl -n numaflow-system delete po e2e-api-pod --ignore-not-found=true
 	cat test/manifests/e2e-api-pod.yaml |  sed 's@quay.io/numaproj/@$(IMAGE_NAMESPACE)/@' | sed 's/:latest/:$(VERSION)/' | kubectl -n numaflow-system apply -f -
 	go generate $(shell find ./test/$* -name '*.go')
 	go test -v -timeout 15m -count 1 --tags test -p 1 ./test/$*
 	$(MAKE) cleanup-e2e
-	
+
+image-restart:
+	$(MAKE) image
+	$(MAKE) delete-numaflow-controller-ux-webhook
```
Comment (whynowy) [inline Makefile:132]: `restart-control-plane-components`

File: Makefile [hunk 17]
```diff
@@ -120,15 +120,21 @@ test-sideinputs-e2e:
 test-%:
 	$(MAKE) cleanup-e2e
 	$(MAKE) image e2eapi-image
-	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=controller-manager,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
-	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-ux,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
-	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-webhook,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
+	$(MAKE) delete-numaflow-controller-ux-webhook
```
Comment (whynowy) [inline Makefile:123]: `restart-control-plane-components`

**Comment by whynowy** [comment]
@KeranYang - please take a look when you get a chance, in case I missed anything.

File: Makefile (hunk 16 above)
Comment (akashjkhamkar) [reply Makefile]: fixing this

**Review by whynowy — APPROVED** [review]

**Comment by whynowy** [comment]
fix the lint [issue](https://github.com/numaproj/numaflow/actions/runs/7383033925/job/20083614685#step:10:16) pls.

**Comment by akashjkhamkar** [comment]
@whynowy 
fixed the lints, can we re-run the workflows again please ?

File: Makefile [hunk 18]
```diff
@@ -120,15 +120,21 @@ test-sideinputs-e2e:
...
-	kubectl -n numaflow-system delete po -lapp.kubernetes.io/component=numaflow-webhook,app.kubernetes.io/part-of=numaflow --ignore-not-found=true
+	$(MAKE) restart-control-plane-components
 	kubectl -n numaflow-system delete po e2e-api-pod --ignore-not-found=true
 	cat test/manifests/e2e-api-pod.yaml |  sed 's@quay.io/numaproj/@$(IMAGE_NAMESPACE)/@' | sed 's/:latest/:$(VERSION)/' | kubectl -n numaflow-system apply -f -
 	go generate $(shell find ./test/$* -name '*.go')
 	go test -v -timeout 15m -count 1 --tags test -p 1 ./test/$*
 	$(MAKE) cleanup-e2e
-	
+
+image-restart:
```
Comment (KeranYang) [inline Makefile:130]: Do we use `image-restart` anywhere? If not, can we remove it? It's a shallow command hence removing should be fine.

File: docs/user-guide/reference/autoscaling.md [hunk 19]
```diff
@@ -53,8 +53,8 @@ spec:
   start processing, because the autoscaling algorithm will divide the TPS by the number of pods even if the pod is not `Running`.
 - `scaleDownCooldownSeconds` - After a scaling operation, how many seconds to wait for the same vertex, if the follow-up
   operation is a scaling down, defaults to `90`.
-- `zeroReplicaSleepSeconds` - How many seconds it will wait after scaling down to `0`, defaults to `120`.
-  Numaflow autoscaler periodically scales up a vertex pod to "peek" the incoming data, this is the period of time to wait before peeking.
+- `zeroReplicaSleepSeconds` - How many seconds it will wait after scaling a source vertex replicas down to `0`, defaults to `120`.
```
Comment (KeranYang) [inline docs/user-guide/reference/autoscaling.md:56]: Please also update the API [here](https://github.com/numaproj/numaflow/blob/ee36df51040ad2a71ac513aa196a126d82381ede/pkg/apis/numaflow/v1alpha1/vertex_types.go#L561). After making the change, please re-build the project which will automatically update the API descriptions, website etc. accordingly.

**Comment by akashjkhamkar** [comment]
`TestComplexSlidingWindowPipeline` is passing on the local

not sure why pods weren't reachable in the CI test run.

Should we try re-running the job ? @whynowy 

### Locally - 
```
=== RUN   TestReduceSuite/TestComplexSlidingWindowPipeline
    reduce_test.go:206: Creating Pipeline complex-sliding-sum
2024/01/02 21:04:42 GET http://127.0.0.1:8378/redis/get-msg-count-contains?pipelineName=complex-sliding-sum&sinkName=sink&targetStr=30
Handling connection for 8378
Handling connection for 8378
2024/01/02 21:04:42 > 200 OK
2024/01/02 21:04:42 > 0
2024/01/02 21:04:47 GET http://127.0.0.1:8378/redis/get-msg-count-contains?pipelineName=complex-sliding-sum&sinkName=sink&targetStr=30
2024/01/02 21:04:47 > 200 OK
2024/01/02 21:04:47 > 1
2024/01/02 21:04:52 GET http://127.0.0.1:8378/redis/get-msg-count-contains?pipelineName=complex-sliding-sum&sinkName=sink&targetStr=60
2024/01/02 21:04:52 > 200 OK
2024/01/02 21:04:52 > 1
2024/01/02 21:04:57 GET http://127.0.0.1:8378/redis/get-msg-count-contains?pipelineName=complex-sliding-sum&sinkName=sink&targetStr=180
2024/01/02 21:04:57 > 200 OK
2024/01/02 21:04:57 > 982
    reduce_test.go:237: Deleting Pipeline complex-sliding-sum
```

**Review by KeranYang — COMMENTED** [review]
How was this change tested?

File: Makefile (hunk 18 above)
Comment (akashjkhamkar) [reply Makefile:130]: not used anywhere, but quite handy for developing and testing changes quickly

**Comment by KeranYang** [comment]
> TestComplexSlidingWindowPipeline is failing

Should be transient, I am triggering a retry. @akashjkhamkar

**Comment by KeranYang** [comment]
@akashjkhamkar Manual testing is sufficient for now. Thanks!

File: docs/user-guide/reference/autoscaling.md (hunk 19 above)
Comment (akashjkhamkar) [reply docs/user-guide/reference/autoscaling.md:56]: updated and ran  `make pre-push -b` to generate docs and api specs

**Review by KeranYang — APPROVED** [review]

**Comment by akashjkhamkar** [comment]
Last re-run was successful 

But for the last commit, another transient failure happened
@KeranYang

**Comment by KeranYang** [comment]
@akashjkhamkar The transient failure might be introduced by this change. I will take a look. Let's not merge this in before we figure it out.

**Comment by akashjkhamkar** [comment]
Just a hunch about the transient failures, 
I saw this code in [when.go](https://github.com/numaproj/numaflow/blob/main/test/fixtures/when.go#L68) , `SendMessageTo`

From the errors, source pod "in" which is a http source in the test, seems like doesn exists (atleast the ip isn't valid, hence the `connect: no route to host`)

as the comment says, this can happen if the map of ip to vertex has a stale mapping,
but not sure if thats the case here, will need to recreate with more logging

```
	// There could be a rare corner case when a previous added pod gets replaced by a new one, making the mapping entry no longer valid.
	// Considering current e2e tests are all lightweight, we assume no such case for now.
	SendMessageTo(w.vertexToPodIpMapping[labelSelector], vertexName, req)
	return w
}

```

File: pkg/reconciler/vertex/scaling/scaling.go [hunk 20]
```diff
@@ -223,6 +215,37 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
 		s.daemonClientsCache.Add(pl.GetDaemonServiceURL(), daemonClient)
 	}
 
+	partitionBufferLengths, partitionAvailableBufferLengths, totalBufferLength, totalCurrentPending, err := getBufferInfos(ctx, key, daemonClient, pl, vertex)
+	if err != nil {
+		err := fmt.Errorf("error while fetching buffer info, %w", err)
+		return err
+	}
```
Comment (KeranYang) [inline pkg/reconciler/vertex/scaling/scaling.go:222]: @akashjkhamkar , before this change, `getBufferInfos` was only invoked when the vertex is NOT a source. Now it gets invoked when the vertex is a source. I wonder if that's the reason why we transiently not able to create the source pod. Could you verify?

Comment (akashjkhamkar) [reply pkg/reconciler/vertex/scaling/scaling.go:222]: `getBufferInfos ` will only be effective, if the vertex has owned buffers (which is true only for non-source)

so i dont think it has to do with getBufferInfos

**Comment by KeranYang** [comment]
@akashjkhamkar , regarding the e2e failure, I decided to treat it as transient and merge this PR. The reasons are: 1. my recent 3 reruns are all successful. 2. the failures we saw were all related to http sending messages from reduce test TestComplexSlidingWindowPipelin, we have http sending messages in other tests too but those didn't fail, hence even if there is an issue, the issue should more likely be related to reduce instead of this PR.

I will keep monitoring future PRs and if e2e failure becomes consistent, I will create a separate issue to track.

**Comment by KeranYang** [comment]
Thanks @akashjkhamkar for making the change!

**Comment by akashjkhamkar** [comment]
thank you @KeranYang and @whynowy for the quick reviews and help 🙌
